File: v3-sandbox/prototype-keys-java/src/main/java/org/hiero/keys/impl/Ed25519PrivateKey.java
package org.hiero.keys.impl;

import org.bouncycastle.asn1.DEROctetString;
import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
import org.bouncycastle.math.ec.rfc8032.Ed25519;
import org.hiero.keys.*;
import org.hiero.keys.io.KeyEncoding;
import org.hiero.keys.io.KeyFormat;
import org.hiero.keys.io.RawFormat;

import java.io.IOException;
import java.util.Arrays;
import java.util.Objects;

import static org.hiero.keys.impl.Oids.ID_ED25519;

public final class Ed25519PrivateKey implements PrivateKey {
    private final byte[] seed32;

    public Ed25519PrivateKey(final byte[] seed32) {
        this.seed32 = Objects.requireNonNull(seed32, "seed32 must not be null");
        if (seed32.length != 32) {
            throw new IllegalArgumentException("Ed25519 private key must be 32 bytes");
        }
    }

    @Override
    public byte[] sign(final byte[] message) {
        Objects.requireNonNull(message, "message must not be null");
        final byte[] sig = new byte[Ed25519.SIGNATURE_SIZE];
        Ed25519.sign(seed32, 0, message, 0, message.length, sig, 0);
        return sig;
    }

    @Override
    public PublicKey createPublicKey() {
        final byte[] pub = new byte[Ed25519.PUBLIC_KEY_SIZE];
        Ed25519.generatePublicKey(seed32, 0, pub, 0);
        return new Ed25519PublicKey(pub);
    }

    @Override
    public byte[] toRawBytes() {
        return Arrays.copyOf(seed32, seed32.length);
    }

    @Override
    public KeyAlgorithm algorithm() {
        return KeyAlgorithm.ED25519;
    }

    @Override
    public KeyType type() {
        return KeyType.PRIVATE;
    }

    @Override
    public byte[] toBytes(final KeyFormat container) {
        Objects.requireNonNull(container, "container must not be null");
        if (!container.supportsType(KeyType.PRIVATE)) {
            throw new IllegalArgumentException("Container does not support private keys: " + container);
        }
        if (container.getRawFormat() != RawFormat.BYTES) {
            throw new IllegalArgumentException("toBytes requires BYTES format: " + container);
        }
        try {
            if (container.encoding() == KeyEncoding.DER) {
                final PrivateKeyInfo pki = new PrivateKeyInfo(new AlgorithmIdentifier(ID_ED25519),
                        new DEROctetString(seed32));
                return pki.getEncoded("DER");
            }
            throw new IllegalArgumentException("Unsupported container/format for Ed25519 private key: " + container);
        } catch (IOException e) {
            throw new RuntimeException("Error encoding Ed25519 key", e);
        }
    }

    @Override
    public String toString(final KeyFormat container) {
        if (container.getRawFormat() != RawFormat.STRING) {
            throw new IllegalArgumentException("Requested String for non-STRING container: " + container);
        }
        if (container == KeyFormat.PKCS8_WITH_PEM) {
            return PemUtil.toPem("PRIVATE KEY", toBytes(KeyFormat.PKCS8_WITH_DER));
        }
        throw new IllegalArgumentException("Unsupported container for toString: " + container);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        if (o instanceof Key k) {
            if (type() != k.type()) return false;
            if (algorithm() != k.algorithm()) return false;
            return Objects.deepEquals(toRawBytes(), k.toRawBytes());
        }
        return false;
    }

    @Override
    public int hashCode() {
        int result = type().hashCode();
        result = 31 * result + algorithm().hashCode();
        result = 31 * result + Arrays.hashCode(toRawBytes());
        return result;
    }
}

File: v3-sandbox/prototype-python/src/hiero_sdk_v3/keys/_impl/ed25519_private_key.py
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey as _CryptoKey
from cryptography.hazmat.primitives.serialization import Encoding, PrivateFormat, NoEncryption

from ..key_algorithm import KeyAlgorithm
from ..key_type import KeyType
from ..io.key_format import KeyFormat
from ..io.raw_format import RawFormat
from ..private_key import PrivateKey


class Ed25519PrivateKey(PrivateKey):
    def __init__(self, key: _CryptoKey) -> None:
        self._key = key

    @classmethod
    def generate(cls) -> "Ed25519PrivateKey":
        return cls(_CryptoKey.generate())

    @classmethod
    def from_raw(cls, raw: bytes) -> "Ed25519PrivateKey":
        if len(raw) != 32:
            raise ValueError(f"Ed25519 private key must be 32 bytes, got {len(raw)}")
        return cls(_CryptoKey.from_private_bytes(raw))

    @property
    def raw_bytes(self) -> bytes:
        return self._key.private_bytes_raw()

    @property
    def algorithm(self) -> KeyAlgorithm:
        return KeyAlgorithm.ED25519

    @property
    def type(self) -> KeyType:
        return KeyType.PRIVATE

    def sign(self, message: bytes) -> bytes:
        return self._key.sign(message)

    def create_public_key(self) -> "Ed25519PublicKey":
        from .ed25519_public_key import Ed25519PublicKey
        return Ed25519PublicKey(self._key.public_key())

    def to_bytes(self, format: KeyFormat) -> bytes:
        if not format.supports_type(KeyType.PRIVATE):
            raise ValueError(f"Format {format} does not support private keys")
        if format.raw_format != RawFormat.BYTES:
            raise ValueError(f"to_bytes requires a BYTES format, got {format}")
        if format == KeyFormat.PKCS8_WITH_DER:
            return self._key.private_bytes(Encoding.DER, PrivateFormat.PKCS8, NoEncryption())  # nosec
        raise ValueError(f"Unsupported format for Ed25519 private key: {format}")

    def to_string(self, format: KeyFormat) -> str:
        if not format.supports_type(KeyType.PRIVATE):
            raise ValueError(f"Format {format} does not support private keys")
        if format.raw_format != RawFormat.STRING:
            raise ValueError(f"to_string requires a STRING format, got {format}")
        if format == KeyFormat.PKCS8_WITH_PEM:
            return self._key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8, NoEncryption()).decode()  # nosec
        raise ValueError(f"Unsupported format for Ed25519 private key: {format}")
